lib/libc/common/include/string.h

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2024, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_LIB_LIBC_COMMON_INCLUDE_STRING_H_
+#define ZEPHYR_LIB_LIBC_COMMON_INCLUDE_STRING_H_
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#if _POSIX_C_SOURCE >= 200809L || defined(_ZEPHYR_SOURCE)
+size_t strnlen(const char *s, size_t maxlen);
```

Comment:
In general I don't think it is correct to provide prototypes to C library functions like this. As we cannot know if some C library is defining this as a macro, or as a static inline, or..

Reply:
> In general I don't think it is correct to provide prototypes to C library functions like this. As we cannot know if some C library is defining this as a macro, or as a static inline, or..

That seems fairly hypothetical (albeit easily accounted for, if necessary).

Reply:
It is certainly not hypothetical. As @keith-packard mentioned https://github.com/zephyrproject-rtos/zephyr/pull/68261#issuecomment-1916020678 it is very normal for C libraries to name the implementation something else, use a common implementation hidden behind macros, fall into compiler builtins, or in general do any type of tricks for one reason or another.

Reply:
> It is certainly not hypothetical.

Is that what is happening here? Does it happen with any of the other C libraries we support?

E.g. @evgeniy-paltsev - does this happen with your toolchain?

So far, I'm just trying to test the hypothesis. If there are any concrete cases, that would make this scenario less hypothetical.